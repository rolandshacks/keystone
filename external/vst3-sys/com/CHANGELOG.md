# Forked as vst3-com 
# 0.1.0 (2020-04-18)

# 0.2.0 (2020-04-03)
### Fixes

- Fixed docs.rs build [#93](https://github.com/microsoft/com-rs/pull/93)
- Support paths when defining super traits [#110](https://github.com/microsoft/com-rs/pull/110)
- Made interface pointers more correct [#125](https://github.com/microsoft/com-rs/pull/125)
- Fix multiple user generated fields not being generated [#132](https://github.com/microsoft/com-rs/pull/132)

### Added

- `Debug` for `IID` [#128](https://github.com/microsoft/com-rs/pull/128)
- Allow more interfaces [#135](https://github.com/microsoft/com-rs/pull/135)

### Changed

- Define IIDs as strings [#107](https://github.com/microsoft/com-rs/pull/107)
- Require unsafe for interface methods [#120](https://github.com/microsoft/com-rs/pull/120)
- Remove winapi as dependency [#122](https://github.com/microsoft/com-rs/pull/122)
- Interface{Ptr,Rc} => Com{Ptr,Rc} [#129](https://github.com/microsoft/com-rs/pull/129)
- Runtime is now done through stand alone functions [#136](https://github.com/microsoft/com-rs/pull/136)


# 0.1.0 (2019-10-01)

Initial release of the `com` crate. 

A one stop shop for all things related to [COM](https://docs.microsoft.com/en-us/windows/win32/com/component-object-model--com--portal) programming in Rust.

This library exposes various macros, structs and functions to the user for both producing and consuming COM components in an idiomatic manner.
