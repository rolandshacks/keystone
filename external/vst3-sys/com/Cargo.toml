[package]
name = "vst3-com"
version = "0.1.0"
authors = ["Microsoft Corp.", "Mike Hilgendorf"]
description = """
Fork of com-rs
"""
edition = "2018"
license = "MIT"
readme = "./README.md"

[dependencies]
vst3-com-macros = { path = "macros" }
