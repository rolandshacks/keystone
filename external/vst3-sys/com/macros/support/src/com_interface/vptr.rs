use super::vtable;
use proc_macro2::{Ident, TokenStream as HelperTokenStream};
use quote::{format_ident, quote};

pub fn generate(interface_ident: &Ident) -> HelperTokenStream {
    let vptr_ident = ident(&interface_ident.to_string());
    let vtable_ident = vtable::ident(&interface_ident.to_string());

    quote!(
        #[allow(missing_docs)]
        pub type #vptr_ident = *const #vtable_ident;
    )
}

pub fn ident(interface_name: &str) -> Ident {
    format_ident!("{}VPtr", interface_name)
}
