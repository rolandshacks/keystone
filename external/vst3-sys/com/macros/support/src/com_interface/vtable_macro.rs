use proc_macro2::{Ident, TokenStream as HelperTokenStream};
use quote::{format_ident, quote};
use syn::{ItemStruct, Type, TypeBareFn};

pub fn generate(item: &ItemStruct) -> HelperTokenStream {
    let vtable_macro = ident(&item.ident);
    let parent_vtable_binding = gen_parent_vtable_binding(item);
    let vtable_functions = gen_vtable_functions(item);
    let initialized_vtable = gen_initialized_vtable(item);
    quote! {
        #[doc(hidden)]
        #[macro_export]
        macro_rules! #vtable_macro {
            ($class:ty, $offset:ty) => {{
                #parent_vtable_binding
                #vtable_functions
                #initialized_vtable
            }};
        }
    }
}

pub fn ident(struct_ident: &Ident) -> Ident {
    format_ident!(
        "{}_gen_vtable",
        crate::utils::camel_to_snake(&struct_ident.to_string().replace("VTable", ""))
    )
}

fn gen_parent_vtable_binding(item: &ItemStruct) -> HelperTokenStream {
    let parent = item.fields.iter().next();
    if let Some(parent) = parent {
        let is_base = parent
            .ident
            .as_ref()
            .map(|i| i.to_string().ends_with("_base"))
            .unwrap_or(false);
        if is_base {
            let type_path = match &parent.ty {
                syn::Type::Path(type_path) => type_path,
                _ => panic!("vtable fields types must be type paths"),
            };
            let qself = type_path
                .qself
                .as_ref()
                .expect("vtable type paths must use associated types");
            let parent_ty = &qself.ty;
            return quote! {
                let parent_vtable = <#parent_ty as vst3_com::ProductionComInterface<$class>>::vtable::<$offset>();
            };
        }
    }

    quote! {}
}

fn gen_vtable_functions(item: &ItemStruct) -> HelperTokenStream {
    let mut functions = Vec::new();
    for field in &item.fields {
        let method_name = field
            .ident
            .as_ref()
            .expect("Only works with structs with named fields");
        match &field.ty {
            Type::Path(_) => {}
            Type::BareFn(fun) => {
                functions.push(gen_vtable_function(&item.ident, method_name, fun));
            }
            _ => panic!("Only supports structs with fields that are functions"),
        };
    }
    quote! {
        #(#functions)*
    }
}

fn gen_vtable_function(
    struct_ident: &Ident,
    method_name: &Ident,
    fun: &TypeBareFn,
) -> HelperTokenStream {
    assert!(fun.unsafety.is_some(), "Function must be marked unsafe");
    assert!(fun.abi.is_some(), "Function must have marked ABI");
    let method_name = format_ident!("{}", crate::utils::camel_to_snake(&method_name.to_string()));
    let interface_name = struct_ident.to_string().replace("VTable", "");
    let interface_ident = format_ident!("{}", interface_name);
    let function_ident = format_ident!("{}_{}", interface_name.to_lowercase(), method_name);
    let params: Vec<_> = fun
        .inputs
        .iter()
        .enumerate()
        .map(|(i, input)| {
            let ident = format_ident!("arg{}", i);
            let ty = &input.ty;
            quote! { #ident: #ty, }
        })
        .collect();
    let args = (0..(params.len())).skip(1).map(|i| {
        let ident = format_ident!("arg{}", i);
        quote! { #ident, }
    });
    let return_type = &fun.output;
    quote! {
        #[allow(missing_docs)]
        unsafe extern "system" fn #function_ident<C: #interface_ident, O: vst3_com::offset::Offset>(#(#params)*) #return_type {
            let this = arg0.sub(O::VALUE) as *const C as *mut C;
            (*this).#method_name(#(#args)*)
        }
    }
}

fn gen_initialized_vtable(item: &ItemStruct) -> HelperTokenStream {
    let name = &item.ident;
    let methods = gen_vtable_method_initialization(item);
    quote! {
        #[allow(missing_docs)]
        #name {
            #methods
        }
    }
}

fn gen_vtable_method_initialization(item: &ItemStruct) -> HelperTokenStream {
    let mut methods = Vec::new();
    for field in &item.fields {
        let method_ident = field
            .ident
            .as_ref()
            .expect("Only works with structs with named fields");

        let function_ident = if method_ident.to_string().ends_with("_base") {
            let parent = format_ident!("parent_vtable",);
            quote! { #parent, }
        } else {
            let function_ident = format_ident!(
                "{}_{}",
                item.ident.to_string().replace("VTable", "").to_lowercase(),
                crate::utils::camel_to_snake(&method_ident.to_string())
            );
            quote! {
                #function_ident::<$class, $offset>,
            }
        };
        let method = quote! {
            #method_ident: #function_ident
        };
        methods.push(method);
    }

    quote!(
        #(#methods)*
    )
}
