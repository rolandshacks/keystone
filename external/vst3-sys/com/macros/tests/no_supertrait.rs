use com::com_interface;

#[com_interface("cc2d05c7-7d20-4ccb-ad75-1e7fb7c77254")]
pub trait LoneInterface {
    fn do_something(&self);
    fn do_other_thing(&self);
}

fn main() {}
