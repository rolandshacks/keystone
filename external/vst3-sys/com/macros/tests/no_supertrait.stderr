error[E0432]: unresolved import `com`
 --> $DIR/no_supertrait.rs:1:5
  |
1 | use com::com_interface;
  |     ^^^ help: a similar path exists: `vst3_com::com`

error: cannot determine resolution for the attribute macro `com_interface`
 --> $DIR/no_supertrait.rs:3:3
  |
3 | #[com_interface("cc2d05c7-7d20-4ccb-ad75-1e7fb7c77254")]
  |   ^^^^^^^^^^^^^
  |
  = note: import resolution is stuck, try simplifying macro imports
